Library/Homebrew/utils.rb

```diff
@@ -770,3 +770,24 @@ def number_readable(number)
   (numstr.size - 3).step(1, -3) { |i| numstr.insert(i, ",") }
   numstr
 end
+
+def truncate_text_to_approximate_size(s, max_bytes, options = {})
+  front_weight = options[:front_weight] || 0.5
+  glue = options[:glue] || "\n[...snip...]\n"
+  return s if s.bytesize <= max_bytes
+
+  if front_weight < 0 || front_weight > 1
+    raise "opts[:front_weight] must be between 0.0 and 1.0"
+  end
+  scale = 0.99 # just a tad small to allow for the glue
+  while true
```

Comment:
I think all of this complication can be avoided by converting the input string to binary encoding, then cutting off the front/back parts at a fixed byte offset dictated by the limits we've chosen, and then re-interpreting the bytes again as UTF-8 (while dropping/replacing invalid byte sequences). This should be pretty similar to [this `test-bot` code](https://github.com/Homebrew/brew/blob/aa747b915a080633e626cd741fa8f74ae6c5c9d8/Library/Homebrew/cmd/test-bot.rb#L1011-L1017). IMO that's a clearer approach.

Another, also simpler, approach might be to split the log into individual lines, count the bytes in a line, and then use that to determine an appropriate number of leading/trailing lines. This assumes that the length of individual lines doesn't get out of hand, but has the benefit that it won't cut in the middle of a line.